### PR TITLE
SmartStore Api wrapped in promises

### PR DESCRIPTION
**Pull request by trooper2013**
Its missing docs and tests (next pr).  @wmathurin @bhariharan please review the api(s)

Summary 
-Added SmartStoreClient wrapper class  for all store creation and deletion. (Can start the promise chain from here)
-SmartStore extension has the Promises namespace with all supported api(s) wrapped in promises.(Can start or continue the promise chain here)

### CODE REVIEW

File: libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFSmartStoreExtensions.swift [hunk 1]
```diff
@@ -0,0 +1,261 @@
+/*
+ SFSmartStoreClient
```
Comment (bhariharan) [inline libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFSmartStoreExtensions.swift]: `SFSmartStoreExtensions`?

File: libs/SalesforceSwiftSDK/SalesforceSwiftSDKTests/SFRestAPITests.swift [hunk 2]
```diff
@@ -28,42 +28,14 @@ import SalesforceSDKCore
...
 @testable import SalesforceSwiftSDK
 
-class SFRestAPITests: XCTestCase {
-    static var testCredentials: SFOAuthCredentials?
-    static var testConfig: TestConfig?
-    static var setupComplete = false
-    
+class SFRestAPITests: SalesforceSwiftSDKBaseTest {
+  
     override class func setUp() {
```
Comment (bhariharan) [inline libs/SalesforceSwiftSDK/SalesforceSwiftSDKTests/SFRestAPITests.swift]: Do we need this? Isn't `super.setUp()` implicit?

File: libs/SalesforceSwiftSDK/SalesforceSwiftSDKTests/SFRestAPITests.swift [hunk 3]
```diff
@@ -28,42 +28,14 @@ import SalesforceSDKCore
...
-                return SalesforceSwiftSDKTests.refreshCredentials(credentials:(SalesforceSwiftSDKTests.testCredentials)!)
-            }.done { userAccount in
-                SFUserAccountManager.sharedInstance().currentUser = userAccount
-                setupComplete = true
-            }.catch { error  in
-                setupComplete = true
-        }
     }
     
     override func setUp() {
```
Comment (bhariharan) [inline libs/SalesforceSwiftSDK/SalesforceSwiftSDKTests/SFRestAPITests.swift]: Do we need this? Isn't `super.setUp()` implicit?

File: libs/SalesforceSwiftSDK/SalesforceSwiftSDKTests/SFRestAPITests.swift [hunk 4]
```diff
@@ -28,42 +28,14 @@ import SalesforceSDKCore
...
         super.setUp()
-        SalesforceSwiftSDKTests.waitForCompletion(maxWaitTime: 5) { () -> Bool in
-            if (SalesforceSwiftSDKTests.setupComplete) {
-                return true
-            }
-            return false
-        }
     }
     
     override func tearDown() {
```
Comment (bhariharan) [inline libs/SalesforceSwiftSDK/SalesforceSwiftSDKTests/SFRestAPITests.swift]: Do we need this? Isn't `super.tearDown()` implicit?

File: libs/SalesforceSwiftSDK/SalesforceSwiftSDKTests/SFRestAPITests.swift [hunk 5]
```diff
@@ -74,7 +46,6 @@ class SFRestAPITests: XCTestCase {
     
     override class func tearDown() {
```
Comment (bhariharan) [inline libs/SalesforceSwiftSDK/SalesforceSwiftSDKTests/SFRestAPITests.swift]: Do we need this? Isn't `super.tearDown()` implicit?

File: libs/SalesforceSwiftSDK/SalesforceSwiftSDKTests/SalesforceSwiftSDKBaseTest.swift [hunk 6]
```diff
@@ -0,0 +1,90 @@
+/*
+ SmartStoreClientTests
```
Comment (bhariharan) [inline libs/SalesforceSwiftSDK/SalesforceSwiftSDKTests/SalesforceSwiftSDKBaseTest.swift]: `SalesforceSwiftSDKBaseTest`?

File: libs/SalesforceSwiftSDK/SalesforceSwiftSDKTests/SalesforceSwiftSDKTests.swift [hunk 7]
```diff
@@ -27,51 +27,18 @@ import SalesforceSDKCore
 import PromiseKit
 @testable import SalesforceSwiftSDK
 
-class SalesforceSwiftSDKTests: XCTestCase {
-    static var testCredentials: SFOAuthCredentials?
-    static var testConfig: TestConfig?
-    static var setupComplete = false
+class SalesforceSwiftSDKTests: SalesforceSwiftSDKBaseTest {
     
     override class func setUp() {
```
Comment (bhariharan) [inline libs/SalesforceSwiftSDK/SalesforceSwiftSDKTests/SalesforceSwiftSDKTests.swift]: Same comments about `setUp()` and `tearDown()` as above in this file as well.

File: libs/SalesforceSwiftSDK/SalesforceSwiftSDKTests/SmartStoreClientTests.swift [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+import Foundation
+import XCTest
+import SalesforceSDKCore
+import PromiseKit
+import SmartStore
+@testable import SalesforceSwiftSDK
+
+class SmartStoreClientTests: SalesforceSwiftSDKBaseTest {
+    
+    override class func setUp() {
```
Comment (bhariharan) [inline libs/SalesforceSwiftSDK/SalesforceSwiftSDKTests/SmartStoreClientTests.swift]: Same comments about `setUp()` and `tearDown()` as above in this file as well.

**Comment by brandonpage** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR, try to keep changes smaller if you can</td>
    </tr>
  </tbody>
</table>

## Current coverage for SalesforceAnalytics.framework is `43.55%`
✅ *No files affecting coverage found*

---
## Current coverage for SalesforceSDKCore.framework is `15.59%`
✅ *No files affecting coverage found*

---
## Current coverage for SmartStore.framework is `61.85%`
✅ *No files affecting coverage found*

---

> Powered by [xcov](https://github.com/nakiostudio/xcov)

<p align="right" data-meta="generated_by_CI-result">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: libs/SalesforceSwiftSDK/SalesforceSwiftSDKTests/SFRestAPITests.swift (hunk 2 above)
Comment (trooper2013) [reply libs/SalesforceSwiftSDK/SalesforceSwiftSDKTests/SFRestAPITests.swift]: @bhariharan  I intentionally left the override related code around due to addl. setup that may be required for individual test classes. I am not done with the tests, anticipate some setup/teardown related changes.

File: libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFSmartStoreExtensions.swift [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+                count = self.api!.count(with: querySpec, error: &error)
+                if let error = error {
+                    resolver.reject(error)
+                }else {
+                    resolver.fulfill(count)
+                }
+            }
+        }
+        
+        public func query(querySpec: SFQuerySpec, pageIndex: UInt) throws -> Promise<[Any]> {
```
Comment (wmathurin) [inline libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFSmartStoreExtensions.swift]: Should we add builders to make constructing SFSoupSpec, SFQuerySpec or SFSoupIndex easier?

Comment (trooper2013) [reply libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFSmartStoreExtensions.swift]: @wmathurin Yes I should have mentioned that it is pending. Do you see other objects that will benefit from the same pattern in the SmartStore?

Comment (wmathurin) [reply libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFSmartStoreExtensions.swift]: I can't think of any objects.

**Pull request merged**